[kernel 0/n] Introduce `bitcoin-chainstate`

dongcarl opened the pull request:
Part of: #24303

This PR introduces an example/demo `bitcoin-chainstate` executable using said library which can print out information about a datadir and take in new blocks on stdin.

Please read the commit messages for more details.

-----

#### You may ask: WTF?! Why is `index/*.cpp`, etc. being linked in?

This PR is meant only to capture the state of dependencies in our consensus engine as of right now. There are many things to decouple from consensus, which will be done in subsequent PRs. Listing the files out right now in `bitcoin_chainstate_SOURCES` is purely to give us a clear picture of the task at hand, it is **not** to say that these dependencies _belongs_ there in any way.

### TODO

1. Clean up `bitcoin-chainstate.cpp`
   It is quite ugly, with a lot of comments I've left for myself, I should clean it up to the best of my abilities (the ugliness of our init/shutdown might be the upper bound on cleanliness here...)

dongcarl committed:
build: Add example bitcoin-chainstate executable

The bitcoin-chainstate executable serves to surface the dependencies
required by a program wishing to use Bitcoin Core's consensus engine as
it is right now.

More broadly, the _SOURCES list serves as a guiding "North Star" for the
libbitcoinkernel project: as we decouple more and more modules of the
codebase from our consensus engine, this _SOURCES list will grow shorter
and shorter. One day, only what is critical to our consensus engine will
remain. Right now, it's "the minimal list of files to link in to even
use our consensus engine".

[META] In a future commit the libbitcoinkernel library will be extracted
       from bitcoin-chainstate, and the libbitcoinkernel library's
       _SOURCES list will be the list that we aim to shrink.

MarcoFalke commented on src/bitcoin-chainstate.cpp:
```diff
@@ -0,0 +1,255 @@
+};
+
+
+int main(int argc, char* argv[]) {
```
What would happen if this function was empty? If it builds, then the changes here are minimal and build-system only and wouldn't need to concern itself with (C++) code yet.

dongcarl replied:
If it's empty, then we would be able to remove all files in `libbitcoinkernel_la_SOURCES` and it would still compile and link. It being not empty is to enforce that we've linked at least the set of files necessary for this functionality.

MarcoFalke commented on src/bitcoin-chainstate.cpp:
```diff
@@ -0,0 +1,255 @@
+                             [](){ return false; });
+    if (rv.has_value()) {
+        std::cerr << "Failed to load Chain state from your datadir." << std::endl;
+        goto epilogue;
```
Wouldn't the code be easier to read by replacing this with `return;` and letting a destructor of an object that lives in this scope take care of the cleanup?

dongcarl replied:
Unfortunately, with our codebase right now if we don't do the shutdown process in a very carefully sequenced way, there will be nullptr dereferences and other undefined behaviour.

ryanofsky replied:
> I don't mean to reorder the sequence. It is just a suggestion to remove `goto`, which in my eyes is impossible to review.

IMO, it is ok to use goto in a limited way to exit a block, and not worse than `break`. This can be cleaner and better than more convoluted alternatives and there are good practices build around it https://www.kernel.org/doc/html/latest/process/coding-style.html#centralized-exiting-of-functions

MarcoFalke replied:
I don't think we should be using the style guide of the kernel (a project written in C) for a project written in C++.

But given that the code here isn't used for anything after compilation (other than to prove that linking works), I think any style is fine here.

dongcarl replied:
Leaving this thread open for the benefit of future reviewers.

ajtowns replied:
Here's a simple way of avoiding goto without (I think) making things less readable: https://github.com/ajtowns/bitcoin/commit/5d7d8de713961a88a40c236346bbfad4ad2906f0

dongcarl replied:
@ajtowns Hmmm good point... Well since I'm going to introduce a `kernel::Context` object anyway, might be good to just move the init/shutdown logic over there and let RAII handle it. Will play around with it, thanks for the hint!!

MarcoFalke commented on src/bitcoin-chainstate.cpp:
```diff
@@ -0,0 +1,255 @@
+        bool new_block;
+        auto sc = std::make_shared<submitblock_StateCatcher>(block.GetHash());
+        RegisterSharedValidationInterface(sc);
+        bool accepted = chainman.ProcessNewBlock(chainparams, blockptr, /* fForceProcessing */ true, /* fNewBlock */ &new_block);
```
wrong argument names?

maflcko reviewed (approved):
Concept ACK. Left some questions.

dongcarl commented:
Pushed 518c876efe...1d9faa8fc7:
- Addressed https://github.com/bitcoin/bitcoin/pull/24304#discussion_r804072712

dongcarl commented:
@MarcoFalke 
> Do you need to add one? What about adding it to the existing nowallet task?

I'm happy to add it to the existing nowallet task if you think it's appropriate! Lmk!

MarcoFalke commented:
> I'm happy to add it to the existing nowallet task if you think it's appropriate! Lmk!

Yeah, I don't think there are any downsides and only upsides to it, right?

dongcarl committed:
ci: Build bitcoin-chainstate

...to make sure that the linker errors that arise from coupling
regressions are caught by CI.

Adding to the "no wallet" ci job as suggested by MarcoFalke.

MarcoFalke commented:
From ci:

```
clientversion.cpp:19:10: fatal error: 'obj/build.h' file not found
#include <obj/build.h>
         ^~~~~~~~~~~~~
1 error generated.

MarcoFalke commented on .cirrus.yml:
```diff
@@ -259,13 +259,13 @@ task:
 
 task:
-  name: '[no wallet] [bionic]'
+  name: '[no wallet, libbitcoinkernel] [bionic]'
```
in the ci commit: pls remove the @MarcoFalke

dongcarl replied:
I removed the "@" from the commit message, is that sufficient?

ryanofsky commented on src/bitcoin-chainstate.cpp:
```diff
@@ -0,0 +1,255 @@
+
+    ChainstateManager chainman;
+
+    auto rv = LoadChainstate(false,
```
In commit "build: Add example bitcoin-chainstate executable" (01840e6161396ed0ab30a122921f67b8dce45499)

The LoadChainstate and VerifyLoadedChainstate calls here would be cleaner with  b7c7f64dc86f0fcdf07cb1e765f7cccc3a3c8897 suggested in https://github.com/bitcoin/bitcoin/pull/23280#pullrequestreview-825684290 and would let this code print specific error messages instead failing generically and discarding the return codes.

ryanofsky commented on src/bitcoin-chainstate.cpp:
```diff
@@ -0,0 +1,255 @@
+#include <iostream> // for cout and shit
```
In commit "build: Add example bitcoin-chainstate executable" (01840e6161396ed0ab30a122921f67b8dce45499)

Should add the commit description "The bitcoin-chainstate executable serves to..." to the top of this file so it is not lost in git history. Or `bitcoin-chainstate` could accept "-help" option and describe its purpose there.

dongcarl replied:
Done as of 0d90128a3b6acab0fd9d1bc39518e2783b55fdf3

ryanofsky commented on src/bitcoin-chainstate.cpp:
```diff
@@ -0,0 +1,255 @@
+#include <scheduler.h>                // for CScheduler
+#include <script/sigcache.h>          // for InitSignatureCache
+
+using node::fReindex;
+using node::LoadChainstate;
+using node::VerifyLoadedChainstate;
```
In commit "build: Add example bitcoin-chainstate executable" (01840e6161396ed0ab30a122921f67b8dce45499)

I think it would be clearer to drop these using statements, and use the full names symbol below. Especially because `node::fIndex` looks like a local variable, not a global variable without the namespace component.

dongcarl replied:
Done as of 9050346f04

ryanofsky commented on src/bitcoin-chainstate.cpp:
```diff
@@ -19,7 +19,7 @@ using node::fReindex;
 using node::VerifyLoadedChainstate;
 
-const extern std::function<std::string(const char*)> G_TRANSLATION_FUN = nullptr;
+__attribute__ ((visibility("default"))) const extern std::function<std::string(const char*)> G_TRANSLATION_FUN = nullptr;
```
In commit "f" (e5579f11b702ad978cd4f672b362e6346c699234)

I don't think `extern` belongs here. Maybe if you drop the `extern` you could drop the visibility attribute as well?

dongcarl replied:
You're right, I've removed it.

However, we still need the visibility attribute for the reasons detailed in the commit message for 7f7f92f81d06b0a3888fad102d3bc09a64483995

ryanofsky reviewed (approved):
Code review ACK e5579f11b702ad978cd4f672b362e6346c699234 with caveat that I mostly focused on the first commit and did not very closely review the build changes. (I think it could be a good idea to put first commit and other commits into two separate PRs, because the reviewers who are best able to give feedback about the code in the first commit are not necessarily the same people who can provide best feedback about the build commits.)

MarcoFalke commented on configure.ac:
```diff
@@ -1622,6 +1630,10 @@ AC_MSG_CHECKING([whether to build bitcoin-util])
 AC_MSG_RESULT($build_bitcoin_util)
 
+AC_MSG_CHECKING([whether to build experimental bitcoin-chainstate])
+AM_CONDITIONAL([BUILD_BITCOIN_CHAINSTATE], [test x$build_bitcoin_chainstate = xyes])
```
nit: remove x and add missing "quotes" like everywhere else?

dongcarl replied:
Done as of 9050346f04d0435ad71d93ba1df4f5c126c0337b

MarcoFalke commented on src/Makefile.am:
```diff
@@ -764,6 +768,102 @@ bitcoin_util_LDADD = \
+# bitcoin-chainstate binary #
+#
+# TODO: Remove blockfilter.cpp and index/blockfilterindex.cpp after merge of
+#       "Improve Indices on pruned nodes via prune blockers" #21726
```
nit: Maybe remove this TODO and either create a review comment (after merge) on 21726, or create a tracking comment in the parent b-cs issue?

dongcarl replied:
Comment removed as of 9050346f04d0435ad71d93ba1df4f5c126c0337b
Added to "Project-wide TODOs": https://github.com/bitcoin/bitcoin/issues/24303

MarcoFalke commented on src/Makefile.am:
```diff
@@ -764,6 +768,102 @@ bitcoin_util_LDADD = \
+  util/thread.cpp \
+  util/threadnames.cpp \
+  util/time.cpp \
+  util/tokenpipe.cpp \
```
question: This means that the "util library" will be compiled twice? Maybe it could make sense to split libutil out from libcommon and make libkernel depend on libutil?

Similarly, the consensus stuff will be compiled twice. Maybe it could make sense to remove consensus stuff from libnode and instead make libnode depend on libkernel?

dongcarl replied:
> This means that the "util library" will be compiled twice?

Yes, as it stands currently at least these few files of the util library will be compiled twice.

> Maybe it could make sense to split libutil out from libcommon and make libkernel depend on libutil?

Unfortunately, there are plenty of `util/*.cpp` that libbitcoinkernel doesn't use, and I wish to prevent any futher coupling in that regard especially since the files under `util/*.cpp` are largely unrelated.

> Maybe it could make sense to remove consensus stuff from libnode and instead make libnode depend on libkernel?

Unfortunately `bitcoin-chainstate` or `libbitcoinkernel` requires quite a lot of `.cpp` files that `libnode` doesn't, and also vice versa (both ends of the venn diagram are non-empty). So there's no clear hierarchy that we can establish without linking in a large number of unnecessary `.cpp` files which I want to prevent.

The following invocation (in a shell that supports process substitution) will print all the `.cpp` that `bitcoin-chainstate`/`libbitcoinkernel` requires but `libnode` doesn't:
```sh
comm -23 <(make --no-print-directory -C src print-bitcoin_chainstate_SOURCES | cut -d= -f2 | tr ' ' '\n' | grep -v '\.h$' | sort -u) \
         <(make --no-print-directory -C src print-libbitcoin_node_a_SOURCES | cut -d= -f2 | tr ' ' '\n' | grep -v '\.h$' | sort -u)
```

```sh
arith_uint256.cpp
bitcoin-chainstate.cpp
chainparamsbase.cpp
chainparams.cpp
clientversion.cpp
coins.cpp
compat/glibcxx_sanity.cpp
compressor.cpp
consensus/merkle.cpp
consensus/tx_check.cpp
core_read.cpp
deploymentinfo.cpp
fs.cpp
hash.cpp
init/common.cpp
key.cpp
logging.cpp
netaddress.cpp
policy/feerate.cpp
policy/policy.cpp
primitives/block.cpp
primitives/transaction.cpp
pubkey.cpp
random.cpp
randomenv.cpp
scheduler.cpp
script/interpreter.cpp
script/script.cpp
script/script_error.cpp
script/standard.cpp
support/cleanse.cpp
support/lockedpool.cpp
sync.cpp
threadinterrupt.cpp
uint256.cpp
util/asmap.cpp
util/bytevectorhash.cpp
util/getuniquepath.cpp
util/hasher.cpp
util/moneystr.cpp
util/rbf.cpp
util/serfloat.cpp
util/settings.cpp
util/strencodings.cpp
util/syscall_sandbox.cpp
util/system.cpp
util/thread.cpp
util/threadnames.cpp
util/time.cpp
util/tokenpipe.cpp
warnings.cpp
```

All the `.cpp` that `libnode` requires but `bitcoin-chainstate`/`libbitcoinkernel` doesn't:
```sh
comm -13 <(make --no-print-directory -C src print-bitcoin_chainstate_SOURCES | cut -d= -f2 | tr ' ' '\n' | grep -v '\.h$' | sort -u) \
         <(make --no-print-directory -C src print-libbitcoin_node_a_SOURCES | cut -d= -f2 | tr ' ' '\n' | grep -v '\.h$' | sort -u)
```

```sh
addrdb.cpp
addrman.cpp
banman.cpp
blockencodings.cpp
httprpc.cpp
httpserver.cpp
i2p.cpp
index/txindex.cpp
init.cpp
mapport.cpp
net.cpp
net_processing.cpp
node/caches.cpp
node/coin.cpp
node/context.cpp
node/interfaces.cpp
node/miner.cpp
node/minisketchwrapper.cpp
node/psbt.cpp
node/transaction.cpp
noui.cpp
rest.cpp
rpc/blockchain.cpp
rpc/mining.cpp
rpc/misc.cpp
rpc/net.cpp
rpc/rawtransaction.cpp
rpc/server.cpp
rpc/server_util.cpp
torcontrol.cpp
txorphanage.cpp
txrequest.cpp
wallet/init.cpp
```

If we perform a similar analysis on `libbitcoinconsensus.la` we'll see that it compiles the `crypto/*.cpp` files again, which can be eliminated by https://github.com/bitcoin/bitcoin/pull/24322/commits/7c2b41e9f81e52cb30efc2d90ebb4b414b97617f, and there's just `script/bitcoinconsensus.cpp` left.

**Potential followup PR:** Extract out a `libbitcoinconensus_internal.la` which doesn't include `script/bitcoinconsensus.cpp` that `libbitcoinkernel` can link against and avoid unnecessary compilation.

```sh
comm -13 <(make --no-print-directory -C src print-bitcoin_chainstate_SOURCES | cut -d= -f2 | tr ' ' '\n' | grep -v '\.h$' | sort -u) \
         <(make --no-print-directory -C src print-libbitcoinconsensus_la_SOURCES | cut -d= -f2 | tr ' ' '\n' | grep -v '\.h$' | sort -u)
```

```sh
crypto/aes.cpp
crypto/chacha20.cpp
crypto/chacha_poly_aead.cpp
crypto/hkdf_sha256_32.cpp
crypto/hmac_sha256.cpp
crypto/hmac_sha512.cpp
crypto/muhash.cpp
crypto/poly1305.cpp
crypto/ripemd160.cpp
crypto/sha1.cpp
crypto/sha256.cpp
crypto/sha256_sse4.cpp
crypto/sha3.cpp
crypto/sha512.cpp
crypto/siphash.cpp
script/bitcoinconsensus.cpp
```

-----

## Appendix 

For reference, here's the `.cpp` comparison for `libcommon`:

All the `.cpp` that `bitcoin-chainstate`/`libbitcoinkernel` requires but `libcommon` doesn't:

```sh
comm -23 <(make --no-print-directory -C src print-bitcoin_chainstate_SOURCES | cut -d= -f2 | tr ' ' '\n' | grep -v '\.h$' | sort -u) \
         <(make --no-print-directory -C src print-libbitcoin_common_a_SOURCES | cut -d= -f2 | tr ' ' '\n' | grep -v '\.h$' | sort -u)
```

```sh
arith_uint256.cpp
bitcoin-chainstate.cpp
blockfilter.cpp
chain.cpp
chainparamsbase.cpp
clientversion.cpp
compat/glibcxx_sanity.cpp
consensus/merkle.cpp
consensus/tx_check.cpp
consensus/tx_verify.cpp
dbwrapper.cpp
deploymentstatus.cpp
flatfile.cpp
fs.cpp
hash.cpp
index/base.cpp
index/blockfilterindex.cpp
index/coinstatsindex.cpp
logging.cpp
node/blockstorage.cpp
node/chainstate.cpp
node/coinstats.cpp
node/ui_interface.cpp
policy/fees.cpp
policy/packages.cpp
policy/rbf.cpp
policy/settings.cpp
pow.cpp
primitives/block.cpp
primitives/transaction.cpp
pubkey.cpp
random.cpp
randomenv.cpp
script/interpreter.cpp
script/script.cpp
script/script_error.cpp
script/sigcache.cpp
shutdown.cpp
signet.cpp
support/cleanse.cpp
support/lockedpool.cpp
sync.cpp
threadinterrupt.cpp
timedata.cpp
txdb.cpp
txmempool.cpp
uint256.cpp
util/asmap.cpp
util/bytevectorhash.cpp
util/getuniquepath.cpp
util/hasher.cpp
util/moneystr.cpp
util/rbf.cpp
util/serfloat.cpp
util/settings.cpp
util/strencodings.cpp
util/syscall_sandbox.cpp
util/system.cpp
util/thread.cpp
util/threadnames.cpp
util/time.cpp
util/tokenpipe.cpp
validation.cpp
validationinterface.cpp
versionbits.cpp
```

All the `.cpp` that `libcommon` requires but `bitcoin-chainstate`/`libbitcoinkernel` doesn't:

```sh
comm -13 <(make --no-print-directory -C src print-bitcoin_chainstate_SOURCES | cut -d= -f2 | tr ' ' '\n' | grep -v '\.h$' | sort -u) \
         <(make --no-print-directory -C src print-libbitcoin_common_a_SOURCES | cut -d= -f2 | tr ' ' '\n' | grep -v '\.h$' | sort -u)
```

```sh
base58.cpp
bech32.cpp
common/bloom.cpp
core_write.cpp
external_signer.cpp
key_io.cpp
merkleblock.cpp
netbase.cpp
net_permissions.cpp
net_types.cpp
outputtype.cpp
protocol.cpp
psbt.cpp
rpc/external_signer.cpp
rpc/rawtransaction_util.cpp
rpc/util.cpp
script/descriptor.cpp
script/sign.cpp
script/signingprovider.cpp
```

dongcarl replied:
Leaving this thread open for the benefit of future reviewers.

maflcko reviewed (approved):
review ACK da9056423079bcbdcfb23133a6063c37469d58b1

ryanofsky reviewed (approved):
Code review ACK 9050346f04d0435ad71d93ba1df4f5c126c0337b. Main difference since last review is build changes have been scaled back to just add the new executable without the new library for now. There are also a few minor tweaks to bitcoin-chainstate, removing `using` statements and simplifying the G_TRANSLATION_FUN declaration

dongcarl commented on src/bitcoin-chainstate.cpp:
```diff
@@ -0,0 +1,255 @@
+
+
+int main(int argc, char* argv[]) {
+    assert(argc == 2);
```
Done as of 0d90128a3b6acab0fd9d1bc39518e2783b55fdf3

MarcoFalke commented on src/bitcoin-chainstate.cpp:
```diff
@@ -0,0 +1,263 @@
+    if (argc != 2) {
+        std::cerr
+            << "Usage: " << argv[0] << " DATADIR" << std::endl
+            << "Display DATADIR information, and process hex-encoded blocks on standard input." << std::endl;
```
Could clarify that this is experimental, for testing only, and expected to break in future versions?

Otherwise someone might accidentally pass the configure option and then open a bug that their mainnet utxo database is corrupted.

dongcarl replied:
Fixed as of: f0f78cb20b49f0a022bf3661538db17eb58ad280

MarcoFalke commented on src/bitcoin-chainstate.cpp:
```diff
@@ -0,0 +1,263 @@
+// The bitcoin-chainstate executable serves to surface the dependencies required
```
missing copyright header?

dongcarl replied:
Fixed as of: f0f78cb20b49f0a022bf3661538db17eb58ad280

MarcoFalke commented:
Can be taken out of draft now?

dongcarl commented:
Pushed 0ad6a3e81265ef2ad5353c4e0f6553955980d2bc:
- Lots of cleanup in `bitcoin-chainstate.cpp`:
  - Added copyright header
  - More information in header comment
  - Removed personal comments
  - Added comments separating the stages of setup
  - Sorted headers
  - Moved `submitblock_StateCatcher` to where it's used so there's more context
  - Addressed some review comments

MarcoFalke commented:
Assigned 24.0, as this missed the 0.23 feature freeze

MarcoFalke commented on src/Makefile.am:
```diff
@@ -764,6 +768,103 @@ bitcoin_util_LDADD = \
+  $(LIBLEVELDB) \
+  $(LIBLEVELDB_SSE42) \
+  $(LIBMEMENV) \
+  $(BOOST_LIBS)
```
Needs rebase after #24301

dongcarl replied:
Done as of 2c03cec2ff8cdbfd5da92bfb507d218e5c6435b0

MarcoFalke commented on src/bitcoin-chainstate.cpp:
```diff
@@ -0,0 +1,259 @@
+// now.
+//
+// DEVELOPER NOTE: Since this is a "demo-only", experimental, etc. executable,
+//                 it may diverge from Bitcoin Core's coding style.
```
I think this can be removed and if there was any reason to use a different coding style for this binary, it can be explained in the pull request, a review comment, or a code comment.

dongcarl replied:
Some discussion here: https://github.com/bitcoin/bitcoin/pull/24304#discussion_r806194408

MarcoFalke commented on src/bitcoin-chainstate.cpp:
```diff
@@ -0,0 +1,259 @@
+
+const std::function<std::string(const char*)> G_TRANSLATION_FUN = nullptr;
+
+int main(int argc, char* argv[]) {
```
clang-format?

dongcarl replied:
Done as of 2c03cec2ff8cdbfd5da92bfb507d218e5c6435b0

maflcko reviewed (approved):
re-ACK 0ad6a3e81265ef2ad5353c4e0f6553955980d2bc 🔟

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK 0ad6a3e81265ef2ad5353c4e0f6553955980d2bc 🔟
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjLOgwAjdvJIwihxHSRaKHl0iVEsfH9gOsvngLNiBtqBBlXbooA0NFaFXMKce5f
v17lIZqfNLCfyIEZFSrgb4cTRAMURDJNTyVg4pFBCEVcmSxmcGniff8v9Y+pbPj7
Z7WYljWzVJpZeuAQyFwiAr9AerIYUZVndLX5X+QKf+JMBqg1Vv7G7SIfgdlUAzl8
+hIPW144xzl6xZRsTJrBp91GdAAY/vyAz+8EfFay82SAJC6ZE9CGtUEzrYyOQ8S/
vD+5ach5xi2FbOQgXYX+49Hqgx+x0VzAhEI4XeUm0NDKaiRlK15oPegyPXjL+obT
wmE+N1AzpeYJ60Wih64XyT7NB3DaMpExoybmMlgPaOyfLZvC+nMlhjffPG5M3sni
Xl1ffLN873bQgYJf8mT/vO/U+PHTD6iJBk9LzhMYTF5ltW94j7/2XCCB3/3nRCFn
2R0jz8wJxovw2a/MDkI3mdfvzzWQeIugGNZie2fs+Vtn5cgyWfdburKq4d2bUEum
MuiehY0A
=PeLI
-----END PGP SIGNATURE-----
```


</details>

ryanofsky reviewed (approved):
Code review ACK 0ad6a3e81265ef2ad5353c4e0f6553955980d2bc with same caveat that mostly focused on first commit, paid less attention to build changes in second commit. Only changes since last review were documentation changes which all looked very good to me.

Some comments about dependencies were lost, but it is probably better not to have the level of detail they were going into here:

```diff
-    // Initialize signatureCache cuz it's used by...
-    //     <- VerifyECDSASignature
-    //     <- CheckECDSASignature
-    //     <- EvalChecksigPreTapscript
-    //     <- EvalScript
-    //     <- VerifyScript
-    //     <- CScriptCheck::()
-    //     <- CheckInputScripts
-    //     <- ConnectBlock
-    //     <- ConnectTip
-    //     <- ActivateBestChainStep
-    //     <- ActivateBestChain
-    //     <- ProcessNewBlock
     InitSignatureCache();
-
-    // Initialize g_scriptExecutionCache{,Hasher} cuz it's used by...
-    //     <- CheckInputScripts
-    //     <- ConnectBlock
-    //     <- ConnectTip
-    //     <- ActivateBestChainStep
-    //     <- ActivateBestChain
-    //     <- ProcessNewBlock
     InitScriptExecutionCache();
```

dongcarl commented on src/bitcoin-chainstate.cpp:
```diff
@@ -0,0 +1,259 @@
+            << "Usage: " << argv[0] << " DATADIR" << std::endl
+            << "Display DATADIR information, and process hex-encoded blocks on standard input." << std::endl
+            << std::endl
+            << "IMPORTANT: THIS EXECUTABLE IS EXPERIMENTAL, FOR TESTING ONLY, AND EXPECTED TO" << std::endl
+            << "           BREAK IN FUTURE VERSIONS. DO NOT USE ON YOUR ACTUAL DATADIR." << std::endl;
```
Marco's idea, and I agree! 😄

ryanofsky reviewed (approved):
Code review ACK 2c03cec2ff8cdbfd5da92bfb507d218e5c6435b0. Just rebase, comments, formatting change since last review

Sjors reviewed (commented):
Did some light testing of 2c03cec2ff8cdbfd5da92bfb507d218e5c6435b0. Very cool! I fed it block 1 and 2 on a fresh datadir and it was happy. I fed it a duplicate block and it exited. I fed it block 4 and it complained it didn't have the previous block. All fine for now I would say.

Can you add a short documentation, e.g. `contrib/KERNEL.md`, that includes a link to #24303, the configure flag and one example incantation how to use this binary? It doesn't have to explain the entire kernel project.

It might also be good to add a basic functional test for the binary. `test/functional/tool_wallet.py` should be a good template. You may need to add something to `good_prefixes_re` depending on how you name the test file.

A skeleton manpage might be nice too, see plumbing example in #20913.

ryanofsky commented on configure.ac:
```diff
@@ -626,6 +627,12 @@ AC_ARG_ENABLE([util-util],
   [build_bitcoin_util=$build_bitcoin_utils])
 
+AC_ARG_ENABLE([experimental-util-chainstate],
+  [AS_HELP_STRING([--enable-experimental-util-chainstate],
```
re: https://github.com/bitcoin/bitcoin/pull/24304#discussion_r808803775

> You may want to document this in the PR description and elsewhere.
> 
> Also should be in the `./configure` summary.

I believe this is only temporarily built as part of our build system and later moves to an examples folder (and  built with cmake!) So that may a reason not to add it to `--configure` help, though I guess it you could have a comment saying the option is temporary and will be removed later. Agree it makes sense to document in PR description and elsewhere (maybe at the top of `bitcoin-chainstate.cpp`) so it's obvious how to build & test this

fanquake replied:
> So that may a reason not to add it to --configure help,

This option is already included in `./configure --help` output, that's done automatically by autotools.

> Also should be in the ./configure summary.

I don't think it needs to be.

ryanofsky reviewed (commented):
> Can you add a short documentation, e.g. `contrib/KERNEL.md`, that includes a link to #24303, the configure flag and one example incantation how to use this binary? It doesn't have to explain the entire kernel project.

You already know this since you commented there, but for anyone else I started writing up some things about libbitcoinkernel design in #24352 (with a lot of "probably" hedging which I'll remove because it was before I talked to Carl and was guessing about some things).

I think existing https://github.com/bitcoin/bitcoin/blob/master/doc/shared-libraries.md will be a good place to document libbitcoinkernel usage when it gets added by the next PR.

> It might also be good to add a basic functional test for the binary. `test/functional/tool_wallet.py` should be a good template. You may need to add something to `good_prefixes_re` depending on how you name the test file.
> 
> A skeleton manpage might be nice too, see plumbing example in #20913.

I think this is only temporarily built as part of our build system, and later moving to an examples folder with a CMake file, so that would be a reason not to do the test, manpage, etc in this style.

fanquake commented:
> I think this is only temporarily built as part of our build system, and later moving to an examples folder with a CMake file, so that would be a reason not to do the test, manpage, etc in this style.

I agree with Russ. I don't really see a reason to add tests, let alone a manpage for this binary.

Sjors commented:
>  later moving to an examples folder with a CMake file

Having untested examples could lead to code rot though.

I could also imagine this `bitcoin-chainstate` binary actually being useful, especially when you add ZMQ support. In a micro-service like architecture you would feed it blocks from another source (like an edge node) and it would spit out ZMQ feeds. All this without the menace of p2p connections and without the need for using RPC communication. But maybe this is a bit too contrived.

ryanofsky commented:
> Having untested examples could lead to code rot though.

Agreed. I definitely think this should be continue to be built by CI at the very least, and maybe a have a little script feeding it a test block.

> I could also imagine this `bitcoin-chainstate` binary actually being useful, especially when you add ZMQ support.

This could be great, but I believe it should be a separate tool cloned / inspired from this one. When you are using a new library it's helpful to have a simple minimal example along with the fancy demo example.

MarcoFalke commented:
Not sure if this needs a manpage and test, since there is already a test for it (see second commit). The goal of the binary right now is to confirm that  the compiler and linker doesn't fail. I don't think that the binary should be considered stable for end users. Also, the interface is mostly a playground, as I understand it, and not the interface that will be shipped when it is final and ready for end users. Thus, adding a test will make it harder to change the interface later on.

maflcko reviewed (approved):
re-ACK 2c03cec2ff8cdbfd5da92bfb507d218e5c6435b0 🏔

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK 2c03cec2ff8cdbfd5da92bfb507d218e5c6435b0 🏔
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjqVQv9GTYBZ644zKBR/8+NgZAZqR/vzNuVLzgPo9UWDexF6M1LZRVpQaoaz831
PxJui5vAF4Eegt6OPhOEzmW4jpd6s5MpURCBL8BzKiyRxe0uT0MPnzufwpZn3iBP
SVpmoAcgG0ZjM9cdY2iJsLqhf47AHqAqh2K8f9gCoObqloHgjIb5fiK5gXowWLDd
B/zlZgmIaxlS/rgyUkpLa/5kIEOONyO56abJ9PhAlXYKm1GIX5+/GgB0Yb/BJEwL
bP843qBy511C07usDLKp3pPMPAFakDFJPJlmE7evSE5DEKcoXVTQGTjoK1JQ/j4t
Pa2bESI48/V7dBOGYDsEd0AdoW3J2Mkhkbzk1lp6Fu+vCQQbEqIVw5F05p3hUmuJ
RyDJROFVjvONophU2MMLOkL7xlfmNQzXY56ONDsd3OhPWG7bX82Kf0XPZ5rDqDGn
9FWDiAf5/mTscazQ0wcwLZlIHX9gx0v0XOZNcLiStfaeAPq4X3U2cEN+2oPHHQtg
8CMkgV6s
=nVUh
-----END PGP SIGNATURE-----
```


</details>

ajtowns commented:
ACK 2c03cec2ff8cdbfd5da92bfb507d218e5c6435b0

ariard commented:
Post-merge ACK 2c03cec
